### PR TITLE
Hidden categories are not showed in suggested categories

### LINKED ISSUE #3640: Don't show hidden categories in suggested categories
**Summary:**
When uploading an image, we ask users to add categories relevant to the image. Currently, the list of categories that we show also includes hidden categories. According to the Commons page about [hidden categories](https://commons.wikimedia.org/wiki/Commons:Categories#Categories_marked_with_%22HIDDENCAT%22):

> Many non-topical categories are marked with __HIDDENCAT__ or {{Hiddencat}} on the category page.

As they're non-topical there's not much use in suggesting those categories. Also, I've observed that hidden categories are generally added automatically to the corresponding pages and are not meant to be manually added by users. Examples:
* [Category:Pages using authority control without Wikidata link - Wikimedia Commons](https://commons.wikimedia.org/wiki/Category:Pages_using_authority_control_without_Wikidata_link)
* [Category:Uses of Wikidata Infobox - Wikimedia Commons](https://commons.wikimedia.org/wiki/Category:Uses_of_Wikidata_Infobox)

There really seem to be [a lot of hidden categories](https://commons.wikimedia.org/wiki/Category:Hidden_categories) in Commons. Not showing the hidden categories would improve the relevancy of the category search results as it would reduce a lot of irrelevant categories, IMO.

### API that could be used
We can identify hidden categories by sending the `acprop=hidden` to the [`allcategories` API](https://www.mediawiki.org/wiki/API:Allcategories). There seems to be a limitation though. The generator version of that API doesn't seems to be considering the `gacprop=hidden` parameter sent to it. Something to look into if we consider doing this.

#### Example API queries (for reference)
* Using `list=allcategories`: https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&list=allcategories&acprefix=pages&aclimit=200&acprop=hidden
* Using `generator=allcategories`: https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&generator=allcategories&gacprefix=pages&gaclimit=200&gacprop=hidden

### PR DESCRIPTION
**Pull request by PaulinaQuintero**
**Description**
Categories marked as not allowed for the user are now hidden
Fixes #3640 

What changes did you make and why?
I added the functionality to hidde categories marked, because before these categories were shown to the user and and supposed to be prohibited

Tested ProdDebug on Nexus 5x and Nexus 4 with API level 30.
image

### CODE REVIEW

**Comment by PaulinaQuintero** [comment]
There are important information missing in the documentation of the APIS, the way to access to important elements of the generator version of the API allcategories is shown with prop=categoryinfo

**Comment by PaulinaQuintero** [comment]
Sorry in advance for all the commits I had some troubles

File: app/src/main/java/fr/free/nrw/commons/category/CategoryInterface.java [hunk 1]
```diff
@@ -33,9 +33,11 @@
      */
     @GET("w/api.php?action=query&format=json&formatversion=2"
             + "&generator=allcategories")
```
Comment (sivaraam) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryInterface.java:35]: For the API to actually return the `categoryinfo` you should do this
```suggestion
            + "&generator=allcategories&prop=categoryinfo")
```

File: app/src/main/java/fr/free/nrw/commons/category/CategoryInterface.java [hunk 2]
```diff
@@ -33,9 +33,11 @@
...
+
     Single<MwQueryResponse> searchCategoriesForPrefix(@Query("gacprefix") String prefix,
                                                           @Query("gaclimit") int itemLimit, @Query("gacoffset") int offset);
 
+
```
Comment (sivaraam) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryInterface.java:40]: Revert these whitespace changes.

File: data-client/src/main/java/org/wikipedia/dataclient/mwapi/MwQueryPage.java [hunk 3]
```diff
@@ -194,6 +210,9 @@ String source() {
         }
     }
 
+
+
+
```
Comment (sivaraam) [inline data-client/src/main/java/org/wikipedia/dataclient/mwapi/MwQueryPage.java:215]: Revert this whitespace change too.

File: data-client/src/main/java/org/wikipedia/dataclient/mwapi/MwQueryPage.java [hunk 4]
```diff
@@ -234,4 +253,4 @@ public boolean hidden() {
         @SuppressWarnings("unused") private List<String> alias;
         @SuppressWarnings("unused") private List<String> label;
     }
-}
+}
```
Comment (sivaraam) [inline data-client/src/main/java/org/wikipedia/dataclient/mwapi/MwQueryPage.java:256]: Revert this removal of the newline too.

File: app/src/main/java/fr/free/nrw/commons/category/CategoryClient.kt [hunk 5]
```diff
@@ -92,8 +92,9 @@ class CategoryClient @Inject constructor(private val categoryInterface: Category
...
             }
             .map {
-                it.map { page -> page.title().replace(CATEGORY_PREFIX, "") }
+                it.filter { page -> !page.categoryInfo().isHidden
+                }.map { page -> page.title().replace(CATEGORY_PREFIX, "") }
             }
     }
 
-}
+}
```
Comment (sivaraam) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryClient.kt:100]: Revert this removal of the newline.

File: data-client/src/main/java/org/wikipedia/search/SearchResults.java [hunk 6]
```diff
@@ -60,5 +60,4 @@ public SearchResults(@NonNull List<MwQueryPage> pages, @NonNull WikiSite wiki,
     @Nullable public Map<String, String> getContinuation() {
         return continuation;
     }
-}
-
+}
```
Comment (sivaraam) [inline data-client/src/main/java/org/wikipedia/search/SearchResults.java:63]: Revert changes to this file.

**Comment by PaulinaQuintero** [comment]
@sivaraam Thank you, I'm going to check it. Could I open a new issue to add the information of the generator version API of allcategories in the documentation? I think it's so important to have this information in the documentation for future contributions. It could be very useful

**Comment by sivaraam** [comment]
> Could I open a new issue to add the information of the generator version API of allcategories in the documentation?

Its not relevant here. You could discuss about this in the [talk page of API:Allcategories](https://www.mediawiki.org/wiki/API_talk:Allcategories) or you could [be bold](https://en.wikipedia.org/wiki/Wikipedia:Be_bold) and edit the [API help page](https://www.mediawiki.org/wiki/API:Allcategories) yourself. Its called a wiki for a reason :)

File: app/src/main/java/fr/free/nrw/commons/category/CategoryClient.kt [hunk 7]
```diff
@@ -92,8 +92,8 @@ class CategoryClient @Inject constructor(private val categoryInterface: Category
                 it.query()?.pages() ?: emptyList()
             }
             .map {
-                it.map { page -> page.title().replace(CATEGORY_PREFIX, "") }
+                it.filter { page -> !page.categoryInfo().isHidden
+                }.map { page -> page.title().replace(CATEGORY_PREFIX, "") }
```
Comment (sivaraam) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryClient.kt:96]: I think we need to be careful here as the `categoryInfo` object wouldn't be present in all the responses that are handled by `responseMapper`. So, we should do this:

```suggestion
                it.filter {
                    page -> page.categoryInfo() == null || !page.categoryInfo().isHidden
                }.map { page -> page.title().replace(CATEGORY_PREFIX, "") }
```

Not handling this is likely a reason for the tests failing in Travis CI.

Comment (sivaraam) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryClient.kt:96]: I think we need to be careful here as the `categoryInfo` object wouldn't be present in all the responses that are handled by `responseMapper`. So, we should do this:

```suggestion
                it.filter {
                    page -> page.categoryInfo() == null || !page.categoryInfo().isHidden
                }.map { page -> page.title().replace(CATEGORY_PREFIX, "") }
```

Not handling this is likely a reason for the tests failing in Travis CI.

File: app/src/main/java/fr/free/nrw/commons/category/CategoryInterface.java [hunk 8]
```diff
@@ -32,10 +32,11 @@
      * @return
      */
     @GET("w/api.php?action=query&format=json&formatversion=2"
-            + "&generator=allcategories")
+            + "&generator=allcategories&prop=categoryinfo")
+
```
Comment (sivaraam) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryInterface.java:36]: Remove this additional new line.

File: app/src/main/java/fr/free/nrw/commons/category/CategoryInterface.java [hunk 9]
```diff
@@ -32,10 +32,11 @@
...
     Single<MwQueryResponse> searchCategoriesForPrefix(@Query("gacprefix") String prefix,
                                                           @Query("gaclimit") int itemLimit, @Query("gacoffset") int offset);
-
+
```
Comment (sivaraam) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryInterface.java:39]: Remove this needless whitespace.

File: data-client/src/main/java/org/wikipedia/search/SearchResults.java [hunk 10]
```diff
@@ -35,7 +35,7 @@ public SearchResults() {
      * @param suggestion a search suggestion to show to the user: "Did you mean ...?"
      */
     public SearchResults(@NonNull List<MwQueryPage> pages, @NonNull WikiSite wiki,
-                         @Nullable Map<String, String> continuation, @Nullable String suggestion) {
+        @Nullable Map<String, String> continuation, @Nullable String suggestion) {
```
Comment (sivaraam) [inline data-client/src/main/java/org/wikipedia/search/SearchResults.java:38]: I don't see any need for this file to be changed. So, revert the changes to this file. You could use the following commands to revert the changes in this file:

```bash
git fetch origin
git checkout origin/master -- data-client/src/main/java/org/wikipedia/search/SearchResults.java
```

Assuming origin points to this repo.

**Comment by sivaraam** [comment]
Merging this PR despite a failed build as the test that failed is unrelated to this PR. For reference, this is the failed test:

```
> Task :app:testBetaDebugUnitTest

fr.free.nrw.commons.nearby.NearbyParentFragmentPresenterTest > testFilterByMarkerTypeMultiSelectCHECKED FAILED

    org.mockito.exceptions.verification.WantedButNotInvoked at NearbyParentFragmentPresenterTest.kt:267

366 tests completed, 1 failed, 2 skipped
```

The same failure is observed in `master` build too, at the time of writing.
